### PR TITLE
[Spec] DSpark support prefill/decode disaggregation

### PR DESCRIPTION
**Pull request by zhangxiaolei123456**
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

Roadmap: https://github.com/sgl-project/sglang/issues/30344

Reference PR: #30513 


This PR adds DSpark support for prefill/decode disaggregation on DeepSeek-V4.

On `main`, DSpark can run in the normal non-disaggregated path, but the PD decode path does not receive the target-model hidden states required to bootstrap the draft-side DSpark state. As a result, decode may enter speculative decoding without valid DSpark `spec_info` / draft KV state.

This PR fixes that gap by transferring DSpark target hidden states from Prefill to Decode through the disaggregation protocol, then injecting them into the decode-side draft KV cache before the first draft step.

The implementation keeps the existing non-disaggregated DSpark behavior unchanged and scopes the new behavior to PD mode.

## What Changed Compared With main

### DSpark Hidden State PD Transfer

- Add a new disaggregation state type: `DSPARK_HIDDEN`.
- Let Decode describe the required DSpark hidden transfer in metadata:
  - target layer ids
  - hidden start offset
  - hidden length
  - decode radix-cache policy
  - PP-aware slice metadata
  - dynamic destination buffer information
- Let Prefill capture target hidden states only on the PP rank that owns the DSpark target layers.
- Transfer DSpark hidden states through Mooncake/NIXL using row-addressed dynamic destination buffers.
- Trim hidden transfer windows according to the Prefill-side cached prefix while preserving absolute token offsets.
- Keep fail-fast checks when required hidden rows are incomplete before transfer.

### Decode-Side Draft Bootstrap

- Assemble received DSpark hidden slices on Decode before committing the transferred request.
- Attach the assembled hidden rows to `prefill_tail_hidden_states`.
- Build DSpark draft input from the Prefill output token as the first decode anchor.
- Inject transferred target hidden states into the decode-side draft KV cache before the first draft step.

### PP-Aware Prefill Support

- Add PP-aware hidden slice metadata so only the PP rank owning the target layers sends DSpark hidden states.
- Propagate DSpark auxiliary hidden tensors through PP proxy outputs.
- Match PP outputs by microbatch id / request identity to avoid FIFO mismatch.
- Snapshot PP batch sequence lengths so delayed PP result processing does not read mutated request state.
- Drain failed PP prefill bootstrap requests consistently across PP ranks.
- Preserve PP admission behavior using side-effect-free resource credit probing before consensus.

### Hidden Buffer and Transfer Resource Management

- Add `DSparkHiddenTransferPlan` to describe row/chunk transfer layouts.
- Add decode-side `DSparkHiddenPagePool` to reuse registered GPU receive buffers.
- Add optional decode hidden receive prewarm:
  - `SGLANG_DSPARK_PD_HIDDEN_RECV_PREWARM_ROWS`
  - `SGLANG_DSPARK_PD_HIDDEN_RECV_PREWARM_PAGES`
- Release Prefill hidden rows as soon as the hidden transfer finishes in the Mooncake worker, instead of waiting for the full KV request success.
- Prefer contiguous hidden row allocation and slice copy on Prefill.
- Use `torch.empty` instead of `torch.zeros` when assembling Decode hidden tensors to avoid unnecessary clearing.

### Metadata and Speculative Decode Integration

- Extend disaggregation metadata buffers to carry DSpark prefill-tail hidden state information.
- Extend `DFlashDraftInputV2` / DSpark draft input to carry:
  - `prefill_tail_hidden_states`
  - `prefill_tail_valid_mask`
  - `prefill_tail_start_positions`
  - `prefill_tail_hidden_projected`
- Add non-padded token count metadata needed by DSpark draft forward batches.
- Ensure Decode radix-cache and Prefill radix-cache policy are aligned for DSpark hidden correctness.

### Streaming Hidden Chunk Transfer and Release Semantics
- Add streaming DSpark hidden transfer for Mooncake so long prompts do not require materializing the full `prompt_len * hidden_width` tensor on Prefill or Decode.
- Split DSpark hidden transfer state from KV request completion:
  - hidden chunk ACK only controls streaming window reuse.
  - hidden request done does not imply KV request success.
  - request success still follows the original KV / metadata success path.
- Release Prefill-side hidden source rows after the hidden request is done, before full KV request completion.
- Release Decode-side hidden receive rows on normal request release / abort paths.
- Serialize Mooncake DSpark hidden chunks per room to preserve chunk ordering under ACK-based flow control.
- Send DSpark hidden-only chunks even when the aligned KV page count is zero.
- Flush pending streaming hidden chunks before writing the next chunk, preventing a later chunk from overwriting an unsent `dspark_hidden_current_*` slot.
- Use offset-based streaming source row writes so multi-chunk hidden transfer preserves the absolute hidden token range.
- Add fail-fast checks for source-window overwrite hazards and invalid streaming hidden ordering.
- Add backend capability boundaries:
  - Mooncake supports streaming DSpark hidden transfer.
  - NIXL / MORI keep the non-streaming/default path and do not enter partially implemented streaming release semantics.

## Why These Changes Are Needed

DSpark speculative decoding depends on target-model hidden states to initialize the draft-side state correctly. In PD mode, Prefill and Decode run in separate processes, so Decode cannot derive those hidden states locally.

The new `DSPARK_HIDDEN` transfer path makes the hidden state an explicit part of the PD protocol, similar to KV/state transfer, but with PP-aware slicing because only specific Prefill PP ranks own the DSpark target layers.

The resource probing and transactional allocation logic are needed to avoid PP rank divergence: all PP ranks must agree on which requests enter the pipeline before any rank performs side-effectful resource allocation.

The dynamic registered GPU receive buffer pool avoids CPU bounce and keeps hidden transfer on the GDR path.

## Compatibility

- Non-disaggregated DSpark behavior is unchanged.
- Existing KV / C128 / auxiliary state transfer paths remain intact.
- DSpark PD hidden transfer is activated only when Decode requests `DSPARK_HIDDEN` metadata.
- Default decode hidden receive prewarm is disabled, so default memory footprint is unchanged.

Current Scope and Future Work

This PR supports DSpark PD hidden-state transfer for Prefill with PP. The implementation is PP-aware: each Prefill PP rank only captures and transfers the DSpark target hidden slice owned by its local layer range.

Mooncake now supports streaming DSpark hidden transfer with hidden/KV/request completion semantics separated. NIXL and MORI do not yet implement streaming hidden release semantics and therefore keep the default non-streaming capability boundary.

Prefill radix cache cannot be enabled alone for DSpark PD today.

DSpark hidden transfer must match the KV transfer window. If Prefill radix cache is enabled but Decode radix cache is disabled, Prefill may skip cached prompt tokens and the corresponding target hidden states will be missing on Decode.

Therefore, DSpark PD currently requires Prefill and Decode radix-cache policies to be consistent. Until DeepSeek-V4 Decode radix cache is supported, Prefill radix cache should stay disabled. #31097

Future work will extend this to Prefill with PP + TP + CP and streaming support for NIXL/MORI. That requires generalizing the hidden metadata from PP-only layer slicing to a full PP/TP/CP layout, including tensor-parallel hidden sharding and context-parallel token-range reconstruction on Decode.

## Accuracy Tests

```
Prefill
SGLANG_PP_LAYER_PARTITION="6,5,5,6,5,5,6,5" SGLANG_DISAGGREGATION_BOOTSTRAP_TIMEOUT=600 NCCL_SOCKET_IFNAME=eth0  NCCL_IB_DISABLE=0 SGLANG_DSV4_FP4_EXPERTS=1 GLOO_SOCKET_IFNAME=eth0 python3 -m sglang.launch_server --model-path /data00/models/DeepSeek-V4-Flash-DSpark --host 0.0.0.0 --port 30000 --trust-remote-code --kv-cache-dtype fp8_e4m3 --mem-fraction-static 0.8 --max-running-requests 64 --chunked-prefill-size 8192 --max-prefill-tokens 16384 --pp-size 8  --attention-backend dsv4 --reasoning-parser deepseek-v4 --tool-call-parser deepseekv4 --disable-overlap-schedule --disable-piecewise-cuda-graph  --disaggregation-mode prefill --disaggregation-transfer-backend mooncake --enable-metrics --disaggregation-ib-device mlx5_1,mlx5_2,mlx5_3,mlx5_4 --moe-runner-backend flashinfer_mxfp4 --disable-radix-cache

Decode
SGLANG_DSPARK_PD_HIDDEN_RECV_POOL_TOKENS=65536 SGLANG_DSV4_FP4_EXPERTS=1 SGLANG_JIT_DEEPGEMM_PRECOMPILE=1 SGLANG_OPT_DEEPGEMM_HC_PRENORM=1 SGLANG_OPT_USE_TILELANG_MHC_PRE=1 GLOO_SOCKET_IFNAME=eth0 NCCL_MIN_NCHANNELS=24 NCCL_IB_QPS_PER_CONNECTION=8 sglang serve --trust-remote-code --model-path /data00/models/DeepSeek-V4-Flash-DSpark --tp 8 --dp-size 8 --enable-dp-attention --cuda-graph-max-bs 32 --max-running-requests 256 --enable-metrics --host 0.0.0.0 --port 30000 --mem-fraction-static 0.85 --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4 --moe-runner-backend flashinfer_mxfp4 --disaggregation-mode decode --disaggregation-ib-device "mlx5_1,mlx5_2,mlx5_3,mlx5_4" --speculative-algo DSPARK --tokenizer-worker-num 8 --enable-dp-lm-head --load-balance-method round_robin --swa-full-tokens-ratio 0.8
```

 Hardward | MMLU | GSM8K | QPQA  | aime25 repeats 16 | status
-- | -- | -- | -- | -- | -- 
H20 | 0.885  |  0.954 | 0.910 | 93.96 | PD

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Speed Tests and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.























































































































































































































<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:x: [Run #29816426168](https://github.com/sgl-project/sglang/actions/runs/29816426168)<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: [Run #29816426138](https://github.com/sgl-project/sglang/actions/runs/29816426138)<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

### CODE REVIEW

**Comment by dongyibo** [comment]
@zhangxiaolei123456  Hello, I noticed that previous PRs supported both PD separation and Deepep. Will this PR also support Deepep?

**Comment by zhangxiaolei123456** [comment]
@dongyibo There are too many changes, and they're too complicated, so I plan to remove DeepEP and create a new PR.

**Comment by dongyibo** [comment]
OK，any plan to support DSpark for GLM-5.2？

----------

该邮件从移动设备发送



--------------原始邮件--------------
发件人："zhangxiaolei ***@***.***&gt;;
发送时间：2026年7月16日(星期四) 晚上11:18
收件人："sgl-project/sglang" ***@***.***&gt;;
抄送："bert.Dy ***@***.***&gt;;"Mention ***@***.***&gt;;
主题：Re: [sgl-project/sglang] [Spec] DSpark support prefill/decode disaggregation (PR #31466)
-----------------------------------

 zhangxiaolei123456 left a comment (sgl-project/sglang#31466)
 
@dongyibo There are too many changes, and they're too complicated, so I plan to remove DeepEP and create a new PR.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by yz-tang** [comment]
@zhangxiaolei123456 I fix a problem when use dsv4-flash-dspark in PD mode, Is it related to your changes? #31513

**Comment by zhangxiaolei123456** [comment]
@yz-tang this PR focus on DSpark hidden state transfer from prefill to decode, your PR may not have noticed this.

File: python/sglang/srt/managers/scheduler.py
Comment (hnyls2002) [inline python/sglang/srt/managers/scheduler.py:1]: Can you move this dspark logic out of `scheduler.py`

**Review by hnyls2002 — COMMENTED** [review]
Seems like a lot of duplicated code hunks. Could you please get a roadmap for your PR? This should have a clear breakdown for each hunk of changes.

**Comment by dongyibo** [comment]
@zhangxiaolei123456  hello~ any plan to merge it? Looking forward to this fea!

**Comment by zhangxiaolei123456** [comment]
@dongyibo this pr transfer hidden state, main branch use transfer draft kv so the main branch can support PD and DSpark，but if you want support PP + PD + DSpark，you can use this pr https://github.com/sgl-project/sglang/pull/33863, if you want PP + PD + CP +DSpark, you can add this pr https://github.com/sgl-project/sglang/pull/33870

**Comment by Eviannn** [comment]
@zhangxiaolei123456 hi, can support split target layers across PP ranks?
